source/geometry_model/two_merged_boxes.cc

```diff
@@ -0,0 +1,523 @@
+                  if (cell->face(2)->vertex(GeometryInfo<dim-1>::vertices_per_cell-1)[dim-1] > height_lith)
+                    cell->face(2)->set_boundary_indicator (2*dim+2);
+                  // reset the rest of the boundary to the default indicator 2
+                  else
+                    cell->face(2)->set_boundary_indicator (2);
+                }
+
+              // set the lithospheric part of the back boundary to indicator 2*dim+3
+              if (cell->face(3)->at_boundary())
+                {
+                  if (cell->face(3)->vertex(GeometryInfo<dim-1>::vertices_per_cell-1)[dim-1] > height_lith)
+                    cell->face(3)->set_boundary_indicator (2*dim+3);
+                  // reset the rest of the boundary to the default indicator 3
+                  else
+                    cell->face(3)->set_boundary_indicator (3);
+                }
+            }
+        }
+    }
+
+    template <int dim>
+    void
+    TwoMergedBoxes<dim>::
+    create_coarse_mesh (parallel::distributed::Triangulation<dim> &total_coarse_grid) const
+    {
+      std::vector<unsigned int> lower_rep_vec(lower_repetitions, lower_repetitions+dim);
+      std::vector<unsigned int> upper_rep_vec(upper_repetitions, upper_repetitions+dim);
+
+      // initialize lower and upper coarse grid with same settings as total_coarse_grid
+      parallel::distributed::Triangulation<dim> lower_coarse_grid(total_coarse_grid.get_communicator());
+      parallel::distributed::Triangulation<dim> upper_coarse_grid(total_coarse_grid.get_communicator());
+
+      // create lower_coarse_grid mesh
+      GridGenerator::subdivided_hyper_rectangle (lower_coarse_grid,
+                                                 lower_rep_vec,
+                                                 lower_box_origin,
+                                                 lower_box_origin+lower_extents,
+                                                 false);
+
+      // create upper_coarse_grid mesh
+      GridGenerator::subdivided_hyper_rectangle (upper_coarse_grid,
+                                                 upper_rep_vec,
+                                                 upper_box_origin,
+                                                 upper_box_origin+upper_extents,
+                                                 false);
+
+      // merge the lower and upper mesh into one total_coarse_grid
+      // now we have at least two cells
+      GridGenerator::merge_triangulations(lower_coarse_grid,
+                                          upper_coarse_grid,
+                                          total_coarse_grid);
+
+
+      // set the names of the boundary indicators
+      // unfortunately, the merge_triangulations does not copy
+      // the boundary indicators of the individual meshes
+      for (typename Triangulation<dim>::active_cell_iterator
+           cell = total_coarse_grid.begin_active();
+           cell != total_coarse_grid.end();
+           ++cell)
```

Comment:
Instead of doing it here by hand, couldn't you just call the function you have above?